cmd/hook.go

```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
+		} else {
+			return nil
+		}
+	}
+
+	lf, err := os.OpenFile("test.log", os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		fail("Internal Server Error", "open log file failed: %v", err)
+	}
+	defer lf.Close()
+
+	if git.CheckGitVersionAtLeast("2.29") != nil {
+		fail("Internal Server Error", "git not support proc-receive.")
+	}
+
+	reader := bufio.NewReader(os.Stdin)
+	repoUser := os.Getenv(models.EnvRepoUsername)
+	repoName := os.Getenv(models.EnvRepoName)
+	pusherID, _ := strconv.ParseInt(os.Getenv(models.EnvPusherID), 10, 64)
+	pusherName := os.Getenv(models.EnvPusherName)
+
+	// 1. Version and features negotiation.
+	// S: PKT-LINE(version=1\0push-options atomic...)
+	// S: flush-pkt
+	// H: PKT-LINE(version=1\0push-options...)
+	// H: flush-pkt
+
+	rs, err := readPktLine(reader)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+	}
+	if rs.Type != pktLineTypeData {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	const VersionHead string = "version=1"
+
+	if !strings.HasPrefix(rs.Data, VersionHead) {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	hasPushOptions := false
+	response := []byte(VersionHead)
+	if strings.Contains(rs.Data, "push-options") {
+		response = append(response, byte(0))
+		response = append(response, []byte("push-options")...)
+		hasPushOptions = true
+	}
+	response = append(response, []byte("\n")...)
+
+	rs, err = readPktLine(reader)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+	}
+	if rs.Type != pktLineTypeFlush {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	err = writePktLine(os.Stdout, pktLineTypeData, response)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+	}
+
+	err = writePktLine(os.Stdout, pktLineTypeFlush, nil)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+	}
+
+	// 2. receive commands from server.
+	// S: PKT-LINE(<old-oid> <new-oid> <ref>)
+	// S: ... ...
+	// S: flush-pkt
+	// # receive push-options
+	// S: PKT-LINE(push-option)
+	// S: ... ...
+	// S: flush-pkt
+	hookOptions := private.HookOptions{
+		UserName: pusherName,
+		UserID:   pusherID,
+	}
+	hookOptions.OldCommitIDs = make([]string, 0, hookBatchSize)
+	hookOptions.NewCommitIDs = make([]string, 0, hookBatchSize)
+	hookOptions.RefFullNames = make([]string, 0, hookBatchSize)
+
+	for {
+		rs, err = readPktLine(reader)
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+		}
+		if rs.Type == pktLineTypeFlush {
+			break
+		}
+		t := strings.SplitN(rs.Data, " ", 3)
+		if len(t) != 3 {
+			continue
+		}
+		hookOptions.OldCommitIDs = append(hookOptions.OldCommitIDs, t[0])
+		hookOptions.NewCommitIDs = append(hookOptions.NewCommitIDs, t[1])
+		hookOptions.RefFullNames = append(hookOptions.RefFullNames, t[2])
```

Comment:
It's better to store the 3 variables in one struct such as "Command", and use a slice of Commands to hold all commands from user.

Reply:
It's ok, but this struct has been used in other hooks, change it will broken other code also ...
In my opion, It not a big problem.

Reply:
yeah this is fine. It's not a Command slice like @jiangjianxiao suggests as we would only have to immediately decompose that slice into the lists that get sent to the server.